license-tool/src/main.rs

```diff
@@ -0,0 +1,378 @@
+    package_name: S,
+    source: P,
+) -> Result<()>
+where
+    S: AsRef<str>,
+    P: AsRef<Path>,
+{
+    let package_name = package_name.as_ref();
+    let package_license = packages_licenses.get(package_name).context(format!(
+        "Couldn't find configuration for package '{}'",
+        package_name
+    ))?;
+    println!("{}", get_paths(package_license, source)?.join(" "));
+    Ok(())
+}
+
+/// Fetches all the licenses for the passed map of package licenses
+async fn fetch_all_licenses<P>(
+    packages_licenses: &HashMap<String, PackageLicense>,
+    dest: P,
+) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    for package_license in packages_licenses.values() {
+        fetch_licenses(package_license, &dest).await?;
+    }
+
+    Ok(())
+}
+
+/// Fetches the licenses in the `PackageLicense` object, and creates a copy of them in `dest`
+async fn fetch_licenses<P>(package_license: &PackageLicense, dest: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    for license in &package_license.licenses {
+        let path: PathBuf = get_path(license, &dest)?.into();
+
+        if path.exists() {
+            // Skip if the file already exists
+            continue;
+        }
+
+        match license.license_url.scheme() {
+            "file" => {
+                let license_url = replace_env_vars(&license.license_url)?;
+                fs::copy(license_url.path(), &path)
+                    .context(format!("Failed to copy file from '{}'", license_url.path()))?;
+            }
+            "http" | "https" => {
+                let content = reqwest::get(license.license_url.clone())
+                    .await
+                    .context(format!(
+                        "Failed to download file from '{}'",
+                        license.license_url.to_string()
+                    ))?
+                    .text()
+                    .await?;
+                let mut dest = File::create(&path).context(format!(
+                    "Failed to create file '{}'",
+                    path.display().to_string()
+                ))?;
+                io::copy(&mut content.as_bytes(), &mut dest).context(format!(
+                    "Failed to copy content to '{}'",
+                    path.display().to_string()
+                ))?;
+            }
+            _ => bail!(
+                "Invalid scheme for '{}', valid options are: ['file://', 'http://', 'https://']",
+                license.license_url
+            ),
+        };
+    }
+    Ok(())
+}
+
+/// Returns a list of paths to the destination files for the licenses
+fn get_paths<P>(package_license: &PackageLicense, dest: P) -> Result<Vec<String>>
+where
+    P: AsRef<Path>,
+{
+    let mut all_paths = Vec::new();
+    for license in &package_license.licenses {
+        all_paths.push(get_path(license, &dest)?);
+    }
+
+    Ok(all_paths)
+}
+
+/// Returns the path to the destination file for the license
+fn get_path<P>(license: &License, base: P) -> Result<String>
```

Comment:
nit: I feel like `get_license_destination` or `get_license_destination` might be a better name for this. Same deal for `get_paths`